### ISSUE TITLE
Coupon Usage Limit Not Working as Expected

### ISSUE THREAD

**Issue by ldr2273** [woocommerce/woocommerce#16422]
Regarding Coupons- I'm finding that the _Usage Limits > Usage Limit Per User_ setting is not functioning as expected. Whenever setting the usage limit per user to 1, for example, I would expect that the coupon would only be able to be used once as detailed (logged in users would be checked by User ID and guests by billing email address.)

However, upon testing this numerous times, I'm seeing very inconsistent results. Some emails I tested allowed me to use the coupon twice, another email three times, and another email only once (as expected.) All times, I tested this in incognito mode, as a guest, and I received drastically different results each time. Since I was testing as a guest, it should've been checking my billing email address used, but it's using something else instead it seems. 

## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. Create a Coupon with the following settings:

![Coupon Settings General](https://cloudup.com/cSjHFGCxFwu+)

![Coupon Settings Usage Restriction](https://cloudup.com/cSjHFGCxFwu+)

![Coupon Settings Usage Limits](https://cloudup.com/ccSae5d_Iui+)

2. Place an order using this coupon with any email of your choice. The order should go through successfully the first time:

![Order Received](https://cloudup.com/cnk8gJpv9kV+)

3. Place a second order using this coupon with a different email address. At this point, I received a _Coupon usage limit has been reached_ error, even though I was using a completely new billing email address:

![Coupon usage limit has been reached](https://cloudup.com/c5pyq4w0mcH+)

4. Place a third & forth order using this coupon but this time with the same email address. For me, I was able to place an order twice using the same email address. Sometimes, up to three times. Other times (using new emails to test), I got rejected as expected. Continue to test various email addresses in this fashion until you see the behavior. 

## Expected/actual behavior

I would expect that as a logged out user, it would use my billing email address to cross-check how many times that coupon was used, but it seems it is not successful in doing so:

![Usage Limit based on billing email](https://cloudup.com/c8txb1WLYD7+)

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>
```
`
### WordPress Environment ###

Home URL: https://latestupdate.mystagingwebsite.com
Site URL: https://latestupdate.mystagingwebsite.com
WC Version: 3.1.1
Log Directory Writable: ✔
WP Version: 4.8.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: nginx
PHP Version: 7.0.21
PHP Post Max Size: 100 MB
PHP Time Limit: 300
PHP Max Input Vars: 6144
cURL Version: 7.54.1
OpenSSL/1.0.1t

SUHOSIN Installed: –
Max Upload Size: 100 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.1.1
WC Database Prefix: wp_
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔
MaxMind GeoIP Database: ✔

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 3.1.1

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: booking (booking)
bundle (bundle)
composite (composite)
external (external)
grouped (grouped)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #4 - /shop/
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My account: #7 - /my-account/

### Theme ###

Name: Storefront
Version: 2.2.4
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –
`
```
</details>

**Comment by mikejolley** [comment]
@ldr2273 In your screenshots judging from the order received page, you were logged in?

Thats likely why you could "use the same email address twice". If you're mixing logged in and logged out users but using same email addresses you definitely will see some kind of overlap.

Sticking to 3.1.1 (since the code is quite different in master) see the following:

- Cart validation https://github.com/woocommerce/woocommerce/blob/3.1.1/includes/class-wc-cart.php#L1676-L1707
- Coupon validation https://github.com/woocommerce/woocommerce/blob/3.1.1/includes/class-wc-coupon.php#L787-L797

In pseudo code the cart logic does the following:

1. Grabs all used email addresses for the coupon (these are stored after placing an order)
2. Checks to see if the user is logged in or not:
    - **If logged in** pulls the user email from the **user account**. This is the email address used during registration.
    - **If logged out** pulls the email address from the billing data. This is the only address we have for this type of user.
3. With the given email, compare to the array of used emails and reject or accept.

The coupon validation logic only looks at user ID. This validation can be done sooner because we have the ID before the user tries to place an order.

After placing an order, ID will be stored for logged in users, and billing email for logged out users.

Even though I doubt real world usage would involve mixing same email addresses between logged in and logged out accounts, we can probably make it slightly less confusing by storing user email as well as ID when logged in. That way, if you're logged out and try to use the email of a user account which already used the coupon, you'll be rejected.

**Comment by mikejolley** [comment]
See https://github.com/woocommerce/woocommerce/pull/16440

**Comment by Kyzcreig** [comment]
I am having the same issue, this is really annoying.

**Comment by tobwinn** [comment]
I'm having the same issue. Is it because of 3.2.1?

**Comment by OzzDizz** [comment]
@tobwinn It appeared in 3.2.0 and is also in 3.2.1, there is a fix in 3.2.2 but if you need to fix it before then (like I did) you can replace the code in class-wc-cart.php with that found under the commit at [17237](https://github.com/woocommerce/woocommerce/pull/17237)

**Comment by nanoanno** [comment]
I'm having the same issue on 3.2.1 ... hopefully we will have an update soon :)

**Comment by redesigned** [comment]
I'm having a similar issue in WooCommerce 3.2.1 with WooCommerce Smart Coupons 3.3.7 but reversed.  Many new users are getting "Coupon usage limit has been reached" when they have not used the coupon previously which is frustrating as this is our new user welcome coupon.

**Comment by JohnnyQQQ** [comment]
Same problem WooCommerce 3.2.1 "Coupon usage limit has been reached" and this is a welcome email!

**Comment by mikejolley** [comment]
Multiple people have already said this will be fixed in 3.2.2 and a fix/PR is above. Next week.
